Replace resize_dim() with set_sizes_and_strides()

NarineK opened the pull request:
Summary: Use set_sizes_and_strides instead of resize_dim.

Differential Revision: D13947867

ezyang commented on aten/src/TH/generic/THTensor.cpp:
```diff
@@ -351,12 +351,17 @@ void THTensor_(select)(THTensor *self, THTensor *src, int dimension, int64_t sli
   {
-    self->set_size(d, self->size(d+1));
-    self->set_stride(d, self->stride(d+1));
+    auto self_size = THTensor_sizeLegacyNoScalars(self, d+1);
```
Hmm, I think it's OK to use `self->size(d+1)` here, instead of the legacy one, because it is not permissible to call select on a 0-dim tensor (the only case when this matters.)

NarineK replied:
I see, thank you @ezyang ! I'll fix it!

NarineK replied:
pushed new changes

ezyang commented on aten/src/TH/generic/THTensor.cpp:
```diff
@@ -351,12 +351,17 @@ void THTensor_(select)(THTensor *self, THTensor *src, int dimension, int64_t sli
   THTensor_(narrow)(self, NULL, dimension, sliceIndex, 1);
+
+  std::vector<int64_t> newSize(self->dim()-1);
+  std::vector<int64_t> newStride(self->dim()-1);
```
You need to explicitly initialize everything `< dimension` now, this is missing.

NarineK replied:
pushed the new changes

ezyang reviewed (changes requested):
some bugs

NarineK committed:
Replace resize_dim() with set_sizes_and_strides() (#16732)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16732

Use set_sizes_and_strides instead of resize_dim with.

Reviewed By: ezyang

Differential Revision: D13947867

fbshipit-source-id: d358509f5217cb7838ef8a583928aa5a0c620e18
